bigquery/storage/apiv1beta2/integration_test.go

```diff
@@ -0,0 +1,313 @@
+			}
+			// Construct append request.
+			var protoSchema *storagepb.ProtoSchema
+			if k == 0 {
+				// first message in stream, construct schema
+				_, descriptor := descriptor.ForMessage(rowSet[0])
+				protoSchema = &storagepb.ProtoSchema{
+					ProtoDescriptor: descriptor,
+				}
+			}
+			req := &storagepb.AppendRowsRequest{
+				WriteStream: writeStream,
+				TraceId:     "integration_test",
+				Rows: &storagepb.AppendRowsRequest_ProtoRows{
+					ProtoRows: &storagepb.AppendRowsRequest_ProtoData{
+						Rows: &storagepb.ProtoRows{
+							SerializedRows: serialized,
+						},
+						WriterSchema: protoSchema,
+					},
+				},
+			}
+			stream.Send(req)
+			reqCount = reqCount + 1
+			totalRows = totalRows + int64(len(rowSet))
+			serialized = nil
+		}
+		stream.CloseSend()
+	}()
+
+	// Monitor updates.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			resp, err := stream.Recv()
+			if err == io.EOF {
+				// expected end of stream.
+				break
+			}
+			if err != nil {
+				t.Errorf("got err on recv: %v", err)
+				break
+			}
+
+			if status := resp.GetError(); status != nil {
+				t.Errorf("Response Error: %v", status)
+			}
+			if result := resp.GetAppendResult(); result != nil {
+				// default stream doesn't provide an offset value, only an empty message.
+				respCount = respCount + 1
+			}
+		}
+	}()
+
+	wg.Wait()
+
+	gotRows, err := queryRowCount(ctx, bqClient, testTable)
+	if err != nil {
+		t.Errorf("failed to get row count: %v", err)
+	}
+
+	if gotRows != totalRows {
+		t.Errorf("query result mismatch, got %v want %v", gotRows, totalRows)
+	}
+
+}
+
+func TestThickWriter(t *testing.T) {
+	setupCtx := context.Background()
+	ctx, _ := context.WithTimeout(context.Background(), 15*time.Second)
+
+	writeClient, bqClient := getClients(setupCtx, t)
+	defer writeClient.Close()
+	defer bqClient.Close()
+
+	dataset, cleanupFunc, err := setupTestDataset(setupCtx, t, bqClient)
+	if err != nil {
+		t.Fatalf("failed to init test dataset: %v", err)
+	}
+	defer cleanupFunc()
+
+	testTable := dataset.Table(tableIDs.New())
+
+	schema := bigquery.Schema{
+		{Name: "name", Type: bigquery.StringFieldType},
+		{Name: "value", Type: bigquery.IntegerFieldType},
+	}
+	if err := testTable.Create(ctx, &bigquery.TableMetadata{Schema: schema}); err != nil {
+		t.Fatalf("couldn't create test table %s: %v", testTable.FullyQualifiedName(), err)
+	}
+	writeStream := fmt.Sprintf("projects/%s/datasets/%s/tables/%s/_default", testTable.ProjectID, testTable.DatasetID, testTable.TableID)
+
+	testData := [][]*testdata.SimpleMessage{
+		[]*testdata.SimpleMessage{
+			{Name: "one", Value: 1},
+			{Name: "two", Value: 2},
+			{Name: "three", Value: 3},
+		},
+		[]*testdata.SimpleMessage{
+			{Name: "four", Value: 1},
+			{Name: "five", Value: 2},
+		},
+	}
+
+	tw, err := NewThickWriter(ctx, writeClient, writeStream)
+	if err != nil {
+		t.Fatalf("failed to create writer: %v", err)
+	}
+	tw.RegisterProto(&testdata.SimpleMessage{})
```

Comment:
Why isn't the proto descriptor part of the NewThickWriter constructor? Can we use `tw` without calling `RegisterProto`?

Reply:
I'm still trying to reason about schema evolution and supporting non-proto serialization using proto reflection.  If we're expecting to change protos, I suspect we'll want an explicit "use this proto" thing, but it could end up being an option.